### DISCUSSION TITLE
How to redirect to www?

### DISCUSSION THREAD

**Discussion by codepilotsf** [basecamp/kamal#1214] [Q&A]
I've got Kamal 2 working great and deploying my app to www.example.com. But if a user directly navigates to example.com without "www", they hit an ugly "connection is not secure" error message. How can I get Kamal Proxy to do what (in my humble opinion) it should already be doing by default?

**Comment by tuladhar** [comment]
By default, Kamal Proxy routes requests only to matching hosts and does not currently support redirection. However, you can achieve this functionality as such:
```
proxy:
  hosts:
  - example.com
  - www.example.com
```

**Comment by codepilotsf** [comment]
Okay so then my app responds on both hosts and I guess I just redirect in my application controller?

**Reply by tuladhar** [reply]
Yes

**Comment by codepilotsf** [comment]
Okay thanks for your help. That's too bad that Kamal Proxy doesn't handle something as basic as a redirect to www.

**Reply by emk-klaay** [reply]
It is open source. Feel free to make a PR to solve it. 👍

**Reply by soheil** [reply]
The biggest d*ck answer I've ever seen. Answers like this makes this project look horrible and I'm befuddled why RoR would use a deployment tool like this.

Redirects are as basic as it gets, if you're not gonna use something like nginx and reinvent your own solution in Go at least have the decency to implement the bare minimum features instead of making condescending remarks to your hijacked users.

**Comment by LiamBateman** [comment]
I've been looking for this as well, kind of like what we would do in Apache .htaccess

If I was to look at adding this in would there be appetite for this feature?

**Comment by viktorianer** [comment]
Redirects shouldn’t be a concern of Kamal 2. These are better handled by DNS, a reverse proxy, or the app itself, depending on the use case.

**Reply by soheil** [reply]
DNS doesn't do redirects! Often the point of a redirect to avoid hitting a specific endpoint in the app to begin with. Finally on your last point can you elaborate how you'd implement a nginx reverse proxy using kamal?!

Literally I've scoured the internet without finding a single example or guide on how to do a redirect or setup nginx/apache with this tool.

**Comment by codepilotsf** [comment]
Of those 3 options, I think option 2 is best. Not all DNS providers offer redirection. Checking the host and redirecting at the app level means every request takes a small performance hit. So that leaves number two— which when using Kamal 2 is kamal-proxy. So I believe this should be the role of kamal-proxy (but it currently doesn’t offer this feature).On Nov 24, 2024, at 10:42 AM, Viktor Schmidt ***@***.***> wrote:﻿
Redirects shouldn’t be a concern of Kamal 2. These are better handled by DNS, a reverse proxy, or the app itself, depending on the use case.

**Reply by soheil** [reply]
Seems like a non-answer just to show he's "responding to issues".

**Comment by dmitry** [comment]
The only way to use `accessories` is with an additional HTTP server. I have used Caddy as it's much easier to deploy and set up in terms of configuration and SSL. It also means you have to deal with TLS by yourself, though Caddy has an almost automatic way to activate it: https://caddyserver.com/docs/automatic-https
I have used this method for 3 servers already, but the problem is that if you have multiple Kamal deployments on one server, you have to support separate Kamal deployment for the Caddy proxy to kamal-proxy, in another way it will not work. It took me a couple of hours to find out all the necessary config details to make it.

**Comment by westoque** [comment]
we like separating the behavior of redirecting from naked domain with the www domain so we just opted on setting up a server that's main purpose is to redirect: `yourdomain.com` -> `www.yourdomain.com`

we have since made a project that's kamal deployable and you can deploy to your own $5 server to just redirect domains. super simple:

https://github.com/narralabs/naked-redirect-server

**Comment by RomanKrasavtsev** [comment]
Here is the final solution to redirect www to the root domain, or vice versa, when deploying with Kamal:

1. Add **one method** as the redirect logic to application_controller.rb:
```ruby
# app/controllers/application_controller.rb
before_action :redirect_host
before_action :redirect_www

private

def redirect_host
  preferred_host = "domain.de"

  if request.host != preferred_host
    redirect_to "#{request.protocol}#{preferred_host}#{request.fullpath}", status: :moved_permanently
  end
end

def redirect_www
  if request.host.start_with?("www.")
    redirect_to "#{request.protocol}#{request.host.sub(/^www\./, '')}#{request.fullpath}", status: :moved_permanently
  end
end
```
This ensures that any requests coming to www.domain.de are redirected to domain.de, or vice versa, with a 301 Moved Permanently status.

2. Ensure both domains are listed in your config/deploy.yml under hosts:

```yaml
# config/deploy.yml
hosts:
  - domain.de
  - www.domain.de
 ```
This way, Kamal will bind to both domains and accept traffic from both domain.de and www.domain.de.

**Note**
- Make sure that your DNS is configured properly and that both domains point to the same server. 
- SSL certificates cover both domains.
- Also, if you’re using a reverse proxy like Nginx in front of your app, the redirect may be better handled at that level for performance.

**Reply by brendon** [reply]
It would be handy to have `kamal-proxy` support canonical domains. It could be as simple as redirecting to the first domain in the `hosts` list if enabled.

In the end though it was easier to use Cloudflare to do this as I had both domains already proxied in there. I used these guides, and these days there are even templates to make it very easy:

https://developers.cloudflare.com/fundamentals/manage-domains/redirect-domain/
https://developers.cloudflare.com/fundamentals/manage-domains/manage-subdomains/

You could also throw Caddy in front of `kamal-proxy` and add some config there. It's only a few lines I think because Caddy has nice clean config compared to say Nginx :)

I have Caddy as an accessory:

```
  caddy:
    image: caddy:2.10
    role: web
    options:
      publish:
        - "80:80"
        - "443:443"
    files:
      - config/caddy/Caddyfile:/etc/caddy/Caddyfile
    directories:
      - data:/data
      - conf:/etc/caddy
```

With a file in my project `config/caddy/Caddyfile`:

```
{
	email someone@somewhere.com

	on_demand_tls {
		ask http://kamal-proxy/check
	}
}

https:// {
	tls {
		on_demand
	}

	reverse_proxy http://kamal-proxy
}
```

This is for on demand ssl certificates which kamal-proxy doesn't support yet, but will soon as there's a PR being actively worked on. The email address in the config is important as it's used when generating the certificates.

I don't publish ports for `kamal-proxy`. You also need an endpoint in your app (in my case it's `/check`) that will return a 200 if Caddy is ok to generate a SSL for that given hostname.

**Reply by brotich** [reply]
> In the end though it was easier to use Cloudflare to do this as I had both domains already proxied in there. I used these guides, and these days there are even templates to make it very easy:

this worked beautifully as I was already using Cloudflare. The templates worked as expected

**Comment by codepilotsf** [comment]
For now, I'm just adding this at the top of my `routes.rb` to redirect from www to apex:

```ruby
  # Redirect www to naked domain
  if Rails.env.production?
    constraints(host: /\Awww\./) do
      match "(*path)", via: :all, to: redirect { |_params, req|
        "https://#{req.host.sub(/\Awww\./, '')}#{req.fullpath}"
      }, status: 301
    end
  end
  ```

**Reply by benzado** [reply]
Thank you for sharing, I used this as a starting point for adding redirection to my app.

In case this is useful to you or (more likely) anyone else passes through here looking for clues like I did:

I briefly thought "hey, is `via: :all` really necessary? Shouldn't I only redirect GET?" Perhaps, but if you don't match _everything_ then the request will fall through and the app will respond. So you either need to redirect everything or generate some other response for other methods. Probably better to redirect everything.

I wanted a more limited rule but wanted to avoid additional configuration, so I set it up like this:

First, I assume you want to enable DNS rebinding protection and so you're going to have something like this in your staging or production configuration:

```ruby
config.hosts = [
  "example.com",
  "www.example.com"
]
```

Near the top of `config/routes.rb` I put:

```ruby
draw :redirect_to_canonical_host
```

which causes Rails to load this:

```ruby
# config/routes/redirect_to_canonical_host.rb
hosts = Rails.configuration.hosts || []
return if hosts.length < 2

canonical_host = hosts[0]
hosts[1..].each do |alternate_host|
  match "(*path)", via: :all,
                   constraints: { host: alternate_host },
                   to: redirect(status: :permanent_redirect) { |_params, req|
                     "#{req.scheme}://#{canonical_host}#{req.fullpath}"
                   }
end
```

The net result is that any requests for `www.example.com` will be redirected to the bare domain, any requests for the bare domain will be handled as usual, and Rails will respond 403 Forbidden to any domains not in the list. If you'd prefer to redirect from the bare domain to www, just reverse the order of the hosts list.

**Comment by o-200** [comment]
hey, i just want to tell that i decided to make a [ruby gem](https://github.com/o-200/rack-url-canonicalizer) for redirecting www prefix in Rails. It also have a few improvements (like validating locales). It was originally created to keep the project simple, without nginx or caddy...

I dont really like to place that logic into ApplicationController or Rails Router, so i just made middleware which is more convenient way (in my opinion).

```ruby
# config/application.rb
#
# by default it removes www. prefix
config.middleware.use Rack::UrlCanonicalizer

# or you can enforce www.
config.middleware.use Rack::UrlCanonicalizer, enforce_www: true
```

Glad to see some comments and suggestions
